feat: add chainify error parser and other changes

koderholic opened the pull request:
This PR does the following

1. Adds an Error Parser for Chainify
2. Updates Wallet Core so that a number of calls to chainify are wrapped using the Chainify Error Parser
3. Updates Wallet Core so that calls to oneInch APIs are wrapped using OneInch Error Parsers
4. Updates Wallet Core so that quote calls to Lifi are wrapped using LifiErrorParser
5. Minor updates to Liquality Error structure
6. Captures Custom Errors(thrown in wallet-core and wallet extension) in error-parser and refactor wallet-core so that Custom Errors are replaced with Liquality Errors.
7. Updates to error reporting

Link to Linear:
https://linear.app/liquality/issue/EXT-36/error-and-exception-handling-error-service-qa-engineer

koderholic committed:
feat: add chainify error parser

koderholic committed:
fix: wrap(with errorparser) some calls to chainify
- getTokenDetails
- getTransactionByHash
- nft.transfer
- wallet.sendTransaction
- swap.updateTransactionFee
- wallet.updateTransactionFee

koderholic committed:
fix: fix yarn.lock related build issue

koderholic committed:
fix: wrap calls to oneinch API using error parser

koderholic committed:
fix: update Liquality errors
- make context required
- save context in error obj
- wrapUserErrorMessage no longer accepts context arg

koderholic committed:
fix: resolve cyclic dependency
- Wallet-core depends on error-parser, the reverse should not be true

koderholic committed:
fix: fix type error

koderholic committed:
fix: wrap(with Error Parser) Lifi quote request

koderholic committed:
style: rename some functions
- wrapAync to wrapAsync
- getParser to getErrorParser

koderholic committed:
refactor: refactor error localization

koderholic committed:
refactor: change how error name is set

koderholic committed:
fix: separate error translations
- translations for walletExtension could be different from mobile

koderholic committed:
fix: add more liquality errors & update imports

koderholic committed:
refactor: replace custom errors in wallet-core
- with Liquality Errors

kraikov commented on packages/error-parser/src/reporters/console.ts:
```diff
@@ -1,4 +1,4 @@
+import { LiqualityError } from '../LiqualityErrors/LiqualityError';
 
 export function reportToConsole(error: LiqualityError) {
   console.log(error);
```
Remove

koderholic replied:
Hey @kraikov why do you suggest to remove?

kraikov replied:
Sorry, I saw a console.log and my immediate reaction was to be removed.

koderholic replied:
Oh got it!

monokh commented on packages/wallet-core/src/store/actions/sendTransaction.ts:
```diff
@@ -43,15 +44,20 @@ export const sendTransaction = async (
-    feeAsset: _feeAsset,
-  });
+  const parser = getErrorParser(ChainifyErrorParser);
+  const tx = (await parser.wrapAsync(
```
Is there any way to make the wrapping of chainify calls seamless? 
It strikes me as verbose and error prone to have to wrap every call individually. 
Instead you could return proxies that handle chainify errors when client is created in `clients.ts`

koderholic replied:
Done

monokh commented on packages/error-parser/src/parsers/wrapCustomError.ts:
```diff
@@ -0,0 +1,8 @@
+import { InternalError } from '../LiqualityErrors';
+
+export function wrapCustomError(customError: any) {
```
this makes more sense as `createCustomError` 
when you read it in usage
`throw wrapCustomError(..)`
vs
`throw createCustomError(...)`

monokh replied:
It would be even better if this function didn't exist and instead `InternalError` took an optional `rawError` in constructor. 

Then usage is even more simple:
`throw new InternalError({ raw: CUSTOM_ERRORS.NotFound.LedgerTransportCreator })`

monokh replied:
by proxy i mean js proxy: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy

koderholic replied:
Done

koderholic committed:
fix: use proxy chainify client
 - errors are parsed in one place

koderholic committed:
refactor: simplify throwing internal custom errors

monokh commented on packages/wallet-core/src/utils/getTransactionByHash.ts:
```diff
@@ -0,0 +1,6 @@
+import { Client } from '@chainify/client';
+import { Transaction } from '@chainify/types';
+
+export async function getTransactionByHash(client: Client, hash: string): Promise<Transaction<any>> {
```
What's the purpose of this extra level of indirection?
Just makes it a bit harder to figure out what the actual call is

Could we remove function and just call directly?

koderholic replied:
Oh! was handy in the initial approach of wrapping each call to chainify to reduce verbosity resulting from the wraps. I'll just remove this now. Thanks for the mention.

koderholic replied:
Done

koderholic committed:
refactor: undo extra level of indirection
- for getTransactionByHash

koderholic committed:
fix: add more liquality error types

koderholic committed:
fix: stringify liqualityError when adding to history

koderholic committed:
fix: update error reporting
- report to discord using chainify httpclient
- reporting is non blocking and fails silently

koderholic committed:
fix: add new custom error to error parser

koderholic committed:
fix: override Error.toString() for liqualityError
- along with helpful util functions

koderholic committed:
fix: update parseError to return error as is if
error is already a liquality error

koderholic committed:
fix: sanitize import path

koderholic committed:
fix: report custom internal errors

koderholic committed:
docs: add Readme for error parser

koderholic committed:
ci: add changeset

koderholic committed:
ci: update changeset

koderholic committed:
fix: use json files for error translations

koderholic committed:
fix: correct name field of Liquality errors

koderholic committed:
fix: add missing else statement

koderholic committed:
refactor: rename function

koderholic committed:
fix: change historyItem error to LiqualityError

koderholic committed:
fix: avoid duplicate error reporting

koderholic committed:
fix: allow non LiqErrors in error report function

koderholic committed:
fix: add stack to jsonified liquality error

koderholic committed:
fix: type fixes

koderholic committed:
fix: update how error is logged to console

koderholic committed:
fix: update translations

koderholic committed:
fix: import fixes

koderholic committed:
fix: import fixes

koderholic committed:
fix: update translations

koderholic committed:
fix: add DuplicateTokenSymbolError

koderholic committed:
fix: add DuplicateTokenSymbolError

koderholic committed:
fix: add DuplicateTokenSymbolError

koderholic committed:
fix: add more liquality errors to error parser
- DappNotConnectedError
- UserDeclinedError

koderholic committed:
fix: update Liquality error structure and copies
- replaced causeKey & suggestionKey with simply translationKey
- combine suggestions and cause into one string

koderholic committed:
fix: delete some liquality errors

koderholic committed:
fix: undo usage of error parser in liqualitySwap

koderholic committed:
fix: update imports

koderholic committed:
fix: rebase

koderholic committed:
ci: update changeset
